File: crates/simcity_sim/src/game/transport/lane_pathfinding.rs
//! Congestion-aware lane edge costs and admissible heuristic helpers.
//!
//! Consumed by the combined RoadLane+Lanelet A* (`lanelet::pathfinding`); the legacy
//! standalone lane A* that lived here has been removed.

use crate::game::map::{MapGrid, TilePos};
use crate::game::roads::RoadKind;
use crate::game::traffic::TrafficOccupancy;
use crate::game::transport::pathfinding::PathfindingConfig;

use super::lane_graph::{LaneGraph, LaneId};

/// Context for congestion-aware lane edge costs.
///
/// Mirrors the road-A* congestion model in `pathfinding/cost.rs`, evaluated per
/// destination lane-tile. `jitter_seed` adds a small deterministic per-trip tie-break
/// so vehicles sharing the same OD pair don't all collapse onto a single corridor
/// (0 = disabled).
pub struct LaneCostCtx<'a> {
    pub grid: &'a MapGrid,
    pub traffic: &'a TrafficOccupancy,
    pub cfg: &'a PathfindingConfig,
    /// Per-trip-instance jitter seed (0 disables tie-break). MUST be drawn fresh per
    /// spawned trip (e.g. `sim_rng.rng.random_range(1..=u64::MAX)`), NOT keyed to the
    /// OD pair. If you key this to origin+destination, every vehicle with the same OD
    /// gets the identical seed → identical route → within-OD spread collapses (Rank-4
    /// regression: one corridor saturates).
    pub jitter_seed: u64,
}

/// Integer edge cost for entering `next_id`. Mirrors `pathfinding::cost::step_cost_for_edge`
/// (speed/desirability base + congestion factor + cost_scale), keyed on the destination
/// lane-tile, plus a deterministic per-trip tie-break jitter.
pub(crate) fn lane_edge_cost(ctx: &LaneCostCtx<'_>, graph: &LaneGraph, next_id: LaneId) -> u32 {
    let Some(lane) = graph.get_lane(next_id) else {
        return 1;
    };
    let kind = lane.kind;

    let speed = kind.speed_limit().max(1.0);
    let capacity = (kind.capacity_per_lane_tile() as f32).max(1.0);
    let desirability = kind.desirability().max(0.1);

    let occupancy = ctx
        .grid
        .idx(lane.pos)
        .and_then(|i| ctx.traffic.per_tick_vehicles.get(i).copied())
        .unwrap_or(0) as f32;
    let congestion = (occupancy / capacity).clamp(0.0, ctx.cfg.congestion_max.max(0.0));

    let base_cost = (1.0 / speed) * (1.0 / desirability);
    let congestion_factor = 1.0 + ctx.cfg.congestion_k * congestion;
    let raw = base_cost * congestion_factor * ctx.cfg.cost_scale.max(1.0);

    let base = raw.max(1.0) as u32;
    base.saturating_add(lane_jitter(ctx.jitter_seed, next_id.0))
}

/// Uncongested per-tile base cost for a road kind: `floor((1/speed)*(1/desirability)*cost_scale)`.
/// This is the floor of `lane_edge_cost` with zero congestion and no jitter. Used to price lanelet
/// internal-path tiles, which have no road kind of their own (they carry `dir == None`).
pub(crate) fn base_tile_cost(kind: RoadKind, cfg: &PathfindingConfig) -> u32 {
    let speed = kind.speed_limit().max(1.0);
    let desirability = kind.desirability().max(0.1);
    ((1.0 / speed) * (1.0 / desirability) * cfg.cost_scale.max(1.0)).max(1.0) as u32
}

/// Deterministic per-edge tie-break, derived from a per-trip-instance seed and a stable node id.
/// Range stays a small fraction of base costs so it only breaks ties, never reroutes around
/// real congestion. Pure integer hashing => no RNG state, fully reproducible.
pub(crate) fn lane_jitter(seed: u64, id: u32) -> u32 {
    if seed == 0 {
        return 0;
    }
    // splitmix64-style mix of (seed, node id).
    let mut z = seed ^ (id as u64).wrapping_mul(0x9E37_79B9_7F4A_7C15);
    z = (z ^ (z >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
    z = (z ^ (z >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
    z ^= z >> 31;
    (z % (LANE_JITTER_RANGE as u64)) as u32
}

/// Max additive tie-break in integer A* units. Base lane costs are
/// `(1/speed)*(1/desirability)*cost_scale` ≈ 25 for TwoLane (cost_scale=1000),
/// so a jitter ceiling of 8 only separates otherwise-equal alternatives.
const LANE_JITTER_RANGE: u32 = 8;

/// Minimum possible per-tile lane edge base cost. Scales the admissible heuristic so A* does not
/// degrade to Dijkstra once turn/lane-change penalties exist on the combined lane+lanelet graph.
/// Global minimum over `RoadKind` of `floor((1/speed_limit) * (1/desirability) * cost_scale)`:
/// SixLane `(1/80) * (1/1.6) * 1000 = 7.81 -> 7`. Penalties and jitter are excluded from the
/// heuristic (added only to real edges), so `h` stays a lower bound on the true cost (admissible).
pub(crate) const MIN_PER_TILE_BASE: u32 = 7;

/// Scaled Manhattan distance heuristic between two tiles. Admissible because every real edge costs
/// at least `MIN_PER_TILE_BASE` and a path of Manhattan length `m` costs at least
/// `m * MIN_PER_TILE_BASE`.
pub(crate) fn heuristic_tiles(a: TilePos, b: TilePos) -> u32 {
    let dx = (a.x - b.x).unsigned_abs();
    let dy = (a.y - b.y).unsigned_abs();
    (dx + dy).saturating_mul(MIN_PER_TILE_BASE)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn heuristic_tiles_is_scaled_manhattan() {
        // Regression pin: the heuristic is Manhattan * MIN_PER_TILE_BASE (not raw dx+dy),
        // so A* over penalty-bearing lane costs does not degrade to Dijkstra.
        let a = TilePos { x: 0, y: 0 };
        let b = TilePos { x: 3, y: 2 };
        assert_eq!(heuristic_tiles(a, b), 5 * MIN_PER_TILE_BASE);
    }
}

File: packages/sim/src/transport/lanePathfinding.ts
// Port of crates/simcity_sim/src/game/transport/lane_pathfinding.rs: congestion-aware lane edge
// costs (f32, as Rust computes them) and the admissible heuristic of the combined lane+lanelet A*.
import type { RoadKind, TilePos } from '../commands';
import type { MapGrid } from '../map/grid';
import { capacityPerLaneTile, roadDesirability, roadSpeedLimit } from '../map/roads';
import type { TrafficOccupancy } from '../traffic/occupancy';
import type { LaneGraph } from './laneGraph';
import type { PathfindingConfig } from './pathfinding';

const f32 = Math.fround;
export const U32_MAX = 0xffff_ffff;
const MASK64 = (1n << 64n) - 1n;

/** Max additive tie-break in integer A* units; base lane costs are ~25 for TwoLane. */
export const LANE_JITTER_RANGE = 8;
/** Minimum per-tile base cost over all road kinds (SixLane: floor(1/80 / 1.6 * 1000) = 7). */
export const MIN_PER_TILE_BASE = 7;

export interface LaneCostCtx {
  readonly grid: MapGrid;
  readonly traffic: TrafficOccupancy;
  readonly cfg: PathfindingConfig;
  /** Per-trip tie-break seed, u64; `0n` disables the jitter. */
  readonly jitterSeed: bigint;
}

/** `x as u32` for an f32: truncates, saturates, NaN becomes 0. */
export function f32ToU32(x: number): number {
  if (!(x > 0)) return 0;
  return x >= U32_MAX ? U32_MAX : Math.trunc(x);
}

export function satAddU32(a: number, b: number): number {
  return Math.min(a + b, U32_MAX);
}

export function satMulU32(a: number, b: number): number {
  return Math.min(a * b, U32_MAX);
}

/** Integer cost of entering lane `nextId`, keyed on its tile's occupancy, plus the per-trip jitter. */
export function laneEdgeCost(ctx: LaneCostCtx, graph: LaneGraph, nextId: number): number {
  const lane = graph.getLane(nextId);
  if (lane === undefined) return 1;
  const kind = lane.kind;

  const speed = f32(Math.max(roadSpeedLimit(kind), 1));
  const capacity = f32(Math.max(capacityPerLaneTile(kind), 1));
  const desirability = f32(Math.max(roadDesirability(kind), f32(0.1)));

  const i = ctx.grid.idx(lane.pos);
  const occupancy = f32(i === undefined ? 0 : (ctx.traffic.perTickVehicles[i] ?? 0));
  const congestion = Math.min(Math.max(f32(occupancy / capacity), 0), f32(Math.max(ctx.cfg.congestionMax, 0)));

  const baseCost = f32(f32(1 / speed) * f32(1 / desirability));
  const congestionFactor = f32(1 + f32(ctx.cfg.congestionK * congestion));
  const raw = f32(f32(baseCost * congestionFactor) * f32(Math.max(ctx.cfg.costScale, 1)));

  return satAddU32(f32ToU32(Math.max(raw, 1)), laneJitter(ctx.jitterSeed, nextId));
}

/** Uncongested per-tile cost of a road kind; prices lanelet internal tiles. */
export function baseTileCost(kind: RoadKind, cfg: PathfindingConfig): number {
  const speed = f32(Math.max(roadSpeedLimit(kind), 1));
  const desirability = f32(Math.max(roadDesirability(kind), f32(0.1)));
  const raw = f32(f32(f32(1 / speed) * f32(1 / desirability)) * f32(Math.max(cfg.costScale, 1)));
  return f32ToU32(Math.max(raw, 1));
}

/** Deterministic per-edge tie-break: a splitmix64-style mix of `(seed, id)`, in `0..LANE_JITTER_RANGE`. */
export function laneJitter(seed: bigint, id: number): number {
  if (seed === 0n) return 0;
  let z = seed ^ ((BigInt(id) * 0x9e37_79b9_7f4a_7c15n) & MASK64);
  z = ((z ^ (z >> 30n)) * 0xbf58_476d_1ce4_e5b9n) & MASK64;
  z = ((z ^ (z >> 27n)) * 0x94d0_49bb_1331_11ebn) & MASK64;
  z ^= z >> 31n;
  return Number(z % BigInt(LANE_JITTER_RANGE));
}

/** Manhattan distance scaled by `MIN_PER_TILE_BASE`, so it stays a lower bound on real edge costs. */
export function heuristicTiles(a: TilePos, b: TilePos): number {
  return satMulU32(Math.abs(a.x - b.x) + Math.abs(a.y - b.y), MIN_PER_TILE_BASE);
}
